plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs

```diff
@@ -113,3 +144,21 @@ handleError recorder action' = do
       logWith recorder Warning msg
       pure $ toIdeResult (Left [])
     Right value -> pure $ toIdeResult (Right value)
+
+-----------------------
+-- helper functions
+-----------------------
+
+getAndIncreaseSemanticTokensId :: Action Text
+getAndIncreaseSemanticTokensId = do
+  ShakeExtras{semanticTokensId} <- getShakeExtras
+  liftIO $ atomically $ do
+    i <- readTVar semanticTokensId
+    modifyTVar' semanticTokensId (+1)
+    return $ T.pack $ show i
+
+getPreviousSemanticTokens :: NormalizedUri -> Action (Maybe SemanticTokens)
+getPreviousSemanticTokens uri = getShakeExtras >>= liftIO . atomically . STM.lookup uri . semanticTokensCache
+
+setSemanticTokens :: NormalizedUri -> SemanticTokens -> Action ()
+setSemanticTokens uri tokens = getShakeExtras >>= liftIO . atomically . STM.insert tokens uri . semanticTokensCache
```

Comment:
I am wondering, since we will keep the SemanticTokens Map around until HLS restarts, could we think of situations where we want to clear this store?
How much additional memory will this take, or is memory usage something we have decided to not care about in this plugin, since it is quite expensive already?

Reply:
It stores the last response of semantic tokens for each file.
It is semanticTokens in the lsp format, main cost are the tokens, should be about 5 UInt(32-bit) for a token.
Say 1000 tokens for a file. Should be around 20kb a file. Suppose we have opened 1000 files, 20mb. (I think it should be fine.)

The main mem usage would be the tons of dependency files that we type checked when computing the semantic tokens.

Reply:
Thank you for this calculation, seems reasonable to me!